ast/annotations_test.go

```diff
@@ -399,14 +399,576 @@ p[v] {v = 2}`,
+# authors:
+# - rule
+# schemas:
+# - input.baz: {"type": "string"}
+# custom:
+#  rule: rule
+p = 1`,
+			},
+			moduleToAnalyze:     "module",
+			ruleOnLineToAnalyze: 44,
+			expected: []AnnotationsRef{
+				{ // Rule annotation is always first
+					Path:     MustParseRef("data.test.p"),
+					Location: &Location{File: "module", Row: 44},
+					Annotations: &Annotations{
+						Scope:         "rule",
+						Title:         "rule",
+						Description:   "rule",
+						Organizations: []string{"rule"},
+						RelatedResources: []*RelatedResourceAnnotation{
+							{
+								Ref: mustParseURL("https://rule"),
+							},
+						},
+						Authors: []*AuthorAnnotation{
+							{
+								Name: "rule",
+							},
+						},
+						Schemas: []*SchemaAnnotation{
+							schemaAnnotationFromMap("input.baz", map[string]interface{}{
+								"type": "string",
+							}),
+						},
+						Custom: map[string]interface{}{
+							"rule": "rule",
+						},
+					},
+				},
+				{
+					Path:     MustParseRef("data.test.p"),
+					Location: &Location{File: "module", Row: 44},
+					Annotations: &Annotations{
+						Scope:         "document",
+						Title:         "doc",
+						Description:   "doc",
+						Organizations: []string{"doc"},
+						RelatedResources: []*RelatedResourceAnnotation{
+							{
+								Ref: mustParseURL("https://doc"),
+							},
+						},
+						Authors: []*AuthorAnnotation{
+							{
+								Name: "doc",
+							},
+						},
+						Schemas: []*SchemaAnnotation{
+							schemaAnnotationFromMap("input.bar", map[string]interface{}{
+								"type": "integer",
+							}),
+						},
+						Custom: map[string]interface{}{
+							"doc": "doc",
+						},
+					},
+				},
+				{
+					Path:     MustParseRef("data.test"),
+					Location: &Location{File: "module", Row: 14},
+					Annotations: &Annotations{
+						Scope:         "package",
+						Title:         "pkg",
+						Description:   "pkg",
+						Organizations: []string{"pkg"},
+						RelatedResources: []*RelatedResourceAnnotation{
+							{
+								Ref: mustParseURL("https://pkg"),
+							},
+						},
+						Authors: []*AuthorAnnotation{
+							{
+								Name: "pkg",
+							},
+						},
+						Schemas: []*SchemaAnnotation{
+							schemaAnnotationFromMap("input.foo", map[string]interface{}{
+								"type": "boolean",
+							}),
+						},
+						Custom: map[string]interface{}{
+							"pkg": "pkg",
+						},
+					},
+				},
+			},
+		},
+		{
+			note: "no annotations on rule",
+			modules: map[string]string{
+				"module": `# METADATA
+# title: pkg
+# description: pkg
+package test
+
+# METADATA
+# scope: document
+# title: doc
+# description: doc
+
+p = 1`,
+			},
+			moduleToAnalyze:     "module",
+			ruleOnLineToAnalyze: 11,
+			expected: []AnnotationsRef{
+				{ // Rule entry is always first, even if no annotations are present
+					Path:        MustParseRef("data.test.p"),
+					Location:    &Location{File: "module", Row: 11},
+					Annotations: nil,
+				},
```

Comment:
I'm curious when a rule has no annotations why do we still show an entry for it ?

Reply:
The reasoning is that if we always let the first element in the chain represent the rule, then you can always get the path to the rule, even if there are no annotations. You can also determine from only the chain that the rule has no annotations itself (of course, you could determine this from looking at the `scope` annotation of the first element, but that requires some more logic; and thinking).
Also, this design decision was made before we introduced `rego.metadata.rule()`, so this promise might carry a bit less impact now. `rego.metadata.rule()` basically acts as a short-form of `rego.metadata.chain()[0].annotations`.